### PR TITLE
Multi-output multi-label classifier

### PR DESCRIPTION
**Pull request by alemelis**
Hi @AlexeyAB 
here's something that may turn useful. Also thanks to @subhasish-ghosh for pointing me in the right direction. 

### What is this?
This PR addresses #3679 and  #3598 and enables the training of networks like this

Imgur

### How?
The classification task is still done by loading image labels from filename. However, now each image is allowed to have more than one label. This is reported in `net.truths` which contains the one-hot-encoding for each image. During training, the ground truth array is used in chunks so that a single label is seen by each softmax layer.

The number of classes defined in the `.data` file is the total one. Hence, if we classify animals (dog, cat, and horse) and weather conditions (rainy and sunny), the total number is 5.

Two new parameters have been added to the `softmax` layer: `classes` and `start_idx`. The former specifies the number of classes the layer is going to predict (this is < than the total number of classes). The latter is an offset to specify from which index in `net.truth` the softmax layer should begin comparing the labels.

The multi-output architecture is done in the `.cfg` through the use of `route` layers (see `multi-output-classifier.cfg` for an example).

### Does it train?

yep
Imgur

### Does it work?

It seems so. I trained a net with two `softmax` output layers. The first classifies on 173 classes, the second only 2. The total number of outputs should be 173+2.

Here is the output from last layer (the `route` used to concatenate the output  from the two `softmax`s) along with the label id number. I'm omitting probabilities between 2 and 170 to keep the post short, but those add up to ~1. Then, probs 173 and 174 are again adding to 1.

```
0 0.0622977
1 0.048337
2 0.0441598
...
170 0.00165623
171 0.00158418
172 0.00165164
173 0.747589
174 0.252411
```

### CODE REVIEW

File: cfg/multi-output-classifier.cfg [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+size=1
+stride=1
+pad=1
+activation=linear
+
+[avgpool]
+
+[softmax]
+groups=1
+classes=173
```
Comment (alemelis) [inline cfg/multi-output-classifier.cfg]: this softmax layer uses the ground truth from the first index onwards, so `start_idx` is not needed

File: cfg/multi-output-classifier.cfg [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+stride=1
+pad=1
+activation=linear
+
+[avgpool]
+
+[softmax]
+groups=1
+classes=2
+start_idx=173
```
Comment (alemelis) [inline cfg/multi-output-classifier.cfg]: here instead you need to specify from which index to start couting

File: src/blas.c [hunk 3]
```diff
@@ -255,11 +255,11 @@ void l1_cpu(int n, float *pred, float *truth, float *delta, float *error)
     }
 }
 
-void softmax_x_ent_cpu(int n, float *pred, float *truth, float *delta, float *error)
+void softmax_x_ent_cpu(int n, float *pred, float *truth, float *delta, float *error, int start_idx)
 {
     int i;
     for(i = 0; i < n; ++i){
-        float t = truth[i];
+        float t = truth[i + start_idx];
```
Comment (alemelis) [inline src/blas.c]: `start_idx` is used to shift forward in the ground truth array (by default `start_idx = 0`

File: src/data.c [hunk 4]
```diff
@@ -492,7 +492,7 @@ void fill_truth(char *path, char **labels, int k, float *truth)
             ++count;
         }
     }
-    if(count != 1) printf("Too many or too few labels: %d, %s\n", count, path);
+    // if(count != 1) printf("Too many or too few labels: %d, %s\n", count, path);
```
Comment (alemelis) [inline src/data.c]: this check is avoided to allow the ground truth to have more than one `1` in it.

Probably this is not the most elegant way to do this

File: src/parser.c [hunk 5]
```diff
@@ -291,7 +291,9 @@ connected_layer parse_connected(list *options, size_params params)
 softmax_layer parse_softmax(list *options, size_params params)
 {
 	int groups = option_find_int_quiet(options, "groups", 1);
-	softmax_layer layer = make_softmax_layer(params.batch, params.inputs, groups);
+    int classes = option_find_int_quiet(options, "classes", params.inputs);
+    int start_idx = option_find_int_quiet(options, "start_idx", 0);
+    softmax_layer layer = make_softmax_layer(params.batch, classes, groups, start_idx);
```
Comment (alemelis) [inline src/parser.c]: the default behavior is the legacy one. Without specifying `classes` and `start_idx` in the config `softmax`, nothing changes

**Comment by erikguo** [comment]
@alemelis ,

I want to try your modification. But I don't how did you create your multi-label ground truth file? 

Now the labels.txt only have label string inside.

Thanx

**Comment by alemelis** [comment]
> @alemelis ,
> 
> I want to try your modification. But I don't how did you create your multi-label ground truth file?
> 
> Now the labels.txt only have label string inside.
> 
> Thanx

hi @erikguo 
you do as for normal classification i.e., the image filename should also contain the class label

let's say you want to classify animals and weather conditions, your images should be labelled as

```
image1_dog_sunny.png
image2_cat_sunny.png
image3_cat_rainy.png
...
```

**Comment by erikguo** [comment]
@alemelis ,

Thank you very much.

I've tried today. But I encountered a problem. The learning rate must be very small, for example, 0.00001.

Do you have same problem?

**Comment by erikguo** [comment]
@alemelis ，

Thank you very much !

**Comment by erikguo** [comment]
@alemelis ,

Did you encounter the problem that training for multi-labels is very hard? I must set the learning rate to very small, for example, 1e5.  Or the loss will go higher and higher to nan.

**Comment by alemelis** [comment]
> @alemelis ,
> 
> Did you encounter the problem that training for multi-labels is very hard? I must set the learning rate to very small, for example, 1e5. Or the loss will go higher and higher to nan.

yes indeed, training is not easy
to avoid loss going NaN, a small learning rate is recommended. I suspect there's a bit of tweaking to do with the optimiser, but I didn't explore that

**Comment by AlexeyAB** [comment]
@alemelis Hi,

* Why multi-output is better than SoftMax Tree (with groups)?
* Why is it better than Multi-label classification with Logistic instead of SoftMax?

**Comment by alemelis** [comment]
Hi @AlexeyAB  

> * Why multi-output is better than SoftMax Tree (with groups)?

Do you mean yolo9000? I couldn't train any model in that way, see #2837

> * Why is it better than Multi-label classification with Logistic instead of SoftMax?

I briefly tried to use a logistic layer but it wasn't clear how to setup the training and I didn't investigate that further. 

However, this is slightly different from multi-label classification as it allows to train branching models. Ideally, you can have different branches (i.e. the feature extractors) for each type of class.

**Comment by isra60** [comment]
Can you give a small tutorial of how to train for custom dataset?

**Comment by erikguo** [comment]
@alemelis 

 In the last [route] layers, which layers should be pointed to? the last two softmax layers or the last two conv layers before softmax layer?

Thanx

**Comment by erikguo** [comment]
@alemelis 

I am using enet-b0.cfg to train my custom data. It's alway go INFINITE after almost 3000 iteration. It's very hard to set learning rate. I already set learning rate to 0.00001

**Comment by alemelis** [comment]
@isra60 @erikguo 

I used the ` cfg/multi-output-classifier.cfg` file provided with this PR.  I didn't do much testing as I'm sure the code can be improved.

setting subdivision to 1 wouldn't work on my computer as the GPU I use isn't able to fit all those images in memory

I used this policy for the lr as this is what pjreddie used in the papers to train classifiers

```
learning_rate=0.01
policy=poly
power=4
```

I noticed also that setting a fixed very small lr (say 0.000001) may work as well

**Comment by isra60** [comment]
@alemis but how do you prepare your dataset? This is image classification right? Not object detection with marked boxes am I right?

**Comment by alemelis** [comment]
> @AleMis but how do you prepare your dataset? This is image classification right? Not object detection with marked boxes am I right?

@isra60 as any other classification training [see here](https://pjreddie.com/darknet/train-cifar/)

in practice, your images will have names like

```
image1_dog_sunny.png
image2_cat_sunny.png
image3_cat_rainy.png
...
```

and in your labels.txt file you will list all the possible labels as

```
dog
cat
...
sunny
rainy
...
```

**Comment by isra60** [comment]
But do you use transfer-learning from an existing dataset?? Or do you train everything from scratch??

Thanks.

**Comment by Germminate** [comment]
> > @alemelis ,
> > I want to try your modification. But I don't how did you create your multi-label ground truth file?
> > Now the labels.txt only have label string inside.
> > Thanx
> 
> hi @erikguo
> you do as for normal classification i.e., the image filename should also contain the class label
> 
> let's say you want to classify animals and weather conditions, your images should be labelled as
> 
> ```
> image1_dog_sunny.png
> image2_cat_sunny.png
> image3_cat_rainy.png
> ...
> ```

@alemelis 

My training returns the following error:
Too many ro too few labels: 2, <path\to\file> 

I have 2 categories:
The first category has 2 classes
The second category has 12 classes

My files are labelled as such:
<number>_<first category class>_<second category class>.jpg

Any idea what went wrong?

**Comment by alemelis** [comment]
> But do you use transfer-learning from an existing dataset?? Or do you train everything from scratch??
> 
> Thanks.

@isra60 
the cfg provided is based on darknet-19 (tiny yolo), hence you can transfer learn from those partial weights

**Comment by Germminate** [comment]
@alemelis 
I have issues compiling:
./src/softmax_layer.c:37:4: error: 'softmax_layer {aka struct layer}' has no member named 'start_idx'
   l.start_idx = start_idx;


I have followed through with all above changes to all files in the src folder.

What is missing?

**Comment by Germminate** [comment]
@alemelis 

I am currently having issues with labels.

I followed the label structure of uniquenumber_label1category_label2category.jpg.

It returns "too many or too few labels: 2, /link/to/image \n label 0: category \n label 1: category".

From my understanding, this indicates that the model only accepts one label. 

Please help.

**Comment by alemelis** [comment]
@Germminate 

when you train, you need to provide the `.data` file with the following information
```
classes=<number of TOTAL classes>
train  =<path/to/train.list>
valid  = <path/to/valid.list> (needed if using -map flag)
labels = <path/to/labels.txt>
backup = <path/to/saving/location/for/weights>
```

the error you had got may indicate that the number of classes in your `labels` are not enough. That file should contain one line for each class. The **total** number of classes comes from the sum of the classes for each classifier branch. 

For example, if you want to classify _dog_ and _cat_ (2 classes) in _rain_, _sun_, and _wind_ (3 classes), your `labels.txt` should be 5 lines long

**Comment by Germminate** [comment]
@alemelis 
Thanks for your response. 

I have already double-checked the .data and labels file, and everything is as per what you have mentioned. 

However, I still get the below:
- model load information -

 Too many or too few labels: 2, /path/to/image
 label 0: name1
 label1: name2

**Comment by kossolax** [comment]
Due to https://github.com/AlexeyAB/darknet/blob/0700930396e1e1eeae267d1caadada79214c6e7d/src/data.c#L490 it can cause error if you have a classname in a directory path.

**Comment by Germminate** [comment]
@kossolax 

Not sure if you are responding to my query, but if you are, my directory path does not contain anything referring to any of the classes I have stated in the labels file.

**Comment by alemelis** [comment]
@Germminate and @kossolax 

I noticed that in my `data/` folder, I have a `labels.list` file containing these paths

```
labels/100_0.png
labels/100_1.png
labels/100_2.png
labels/100_3.png
labels/100_4.png
labels/100_5.png
labels/100_6.png
labels/100_7.png
labels/101_0.png
labels/101_1.png
...
```

I do not recall creating it, but it is not in the current master, so it may be needed (@AlexeyAB ?)

**Comment by Germminate** [comment]
> @Germminate and @kossolax
> 
> I noticed that in my `data/` folder, I have a `labels.list` file containing these paths
> 
> ```
> labels/100_0.png
> labels/100_1.png
> labels/100_2.png
> labels/100_3.png
> labels/100_4.png
> labels/100_5.png
> labels/100_6.png
> labels/100_7.png
> labels/101_0.png
> labels/101_1.png
> ...
> ```
> 
> I do not recall creating it, but it is not in the current master, so it may be needed (@AlexeyAB ?)

If I understanding this correctly, 100 and 101 are your classes? If so, labels.list is a list of all the images that I have, is that correct?

Also, how do I include this in the .data file? labels.txt contains the names of the classes that I want the labels to identify. 

many thanks!

File: src/data.c (hunk 4 above)
Comment (kossolax) [reply src/data.c]: This line should be commented instead https://github.com/AlexeyAB/darknet/blame/master/src/data.c#L537 since https://github.com/AlexeyAB/darknet/commit/2a873f34485c75d44a346f92ba7dcf2e2aa57a15

**Comment by kossolax** [comment]
I can't get this working. I've attached a debuger to check if the thurth/inputs/output from two softmax and they are good. The route layer does the correct jobs too. I guess the SSE is ok for the multi-classes...  I've tried noloss to both softmax layer, but it doesn't work too.


Playing with learning rate just report the problem to happens later.

**Comment by kossolax** [comment]
Trained with [yolo] layer instead as multiclass input instead of two softmax for 15K iterations and loss is fine.

**Comment by canyilmaz90** [comment]
@kossolax Could you tell me how did you use [yolo] layer in a classification network? Can you share your .cfg file?

**Comment by kossolax** [comment]
Just like a normal detector (with 1 anchor), but with only one box with
contains the entire image.

**Comment by alemelis** [comment]
> Trained with [yolo] layer instead as multiclass input instead of two softmax for 15K iterations and loss is fine.

hi there!
not sure what you mean with using yolo layer for classification as the yolo layer loss is computed against class, x, y, w, and h values. My guess is that during your training, the network is simply matching all the images to 0, 0, 0, 0, 0 or whatever is the default value for them

**Comment by canyilmaz90** [comment]
@kossolax If so, how do you prepare labels? Is it same with the imagenet  classification, or object detection type which is something like:
```
dog  0.5  0.5  1.0  1.0
sunny  0.5  0.5  1.0  1.0
```

**Comment by kossolax** [comment]
Exactly, yes it is

Le lun. 17 févr. 2020 à 15:33, Can Yılmaz <notifications@github.com> a
écrit :

**Comment by alemelis** [comment]
in that case, it is simply multi-label, not also multi-output

multi-label is already supported by v3

**Comment by canyilmaz90** [comment]
So, for this case, it could return less or more attribute than the image actually have, right?

**Comment by alemelis** [comment]
> So, for this case, it could return less or more attribute than the image actually have, right?

multi-label only would return `dog` and/or `sunny`, or nothing

multi-label multi-output would return `dog-sunny` or nothing

**Comment by canyilmaz90** [comment]
- Can someone explain how the 'groups' parameter in the softmax layer works? (Which is stated by @AlexeyAB above)
- Does this branch work properly with the up-to-date master branch? @alemelis  

Thanks.

**Comment by alemelis** [comment]
>     * Can someone explain how the 'groups' parameter in the softmax layer works? (Which is stated by @AlexeyAB above)
> 
>     * Does this branch work properly with the up-to-date master branch? @alemelis
> 
> 
> Thanks.

@canyilmaz90 it's probably working fine with the master, as it's using a very old part of the codebase, there are some conflicts to resolve tho

**Comment by canyilmaz90** [comment]
Thanks

File: src/data.c (hunk 4 above)
Comment (canyilmaz90) [reply src/data.c]: @kossolax can you clarify where we should comment? Line 537 points out a bracket only

Comment (kossolax) [reply src/data.c]: "Too many or too few labels" in the "fill_truth_smooth" function.

Comment (canyilmaz90) [reply src/data.c]: I guess lines 539-548, which is all of `if (count != 1)` block.

Comment (kossolax) [reply src/data.c]: yes

**Comment by canyilmaz90** [comment]
@alemelis what is the command to test this on images? `./darknet classifier predict` returns only one output

**Comment by xjohnxjohn** [comment]
> @alemelis what is the command to test this on images? `./darknet classifier predict` returns only one output

are you received or finished the answer ?

**Comment by alemelis** [comment]
> @alemelis what is the command to test this on images? `./darknet classifier predict` returns only one output

@canyilmaz90 if you want the output like in the PR notes, you need to remove the last `[cost]` layer from the cfg

**Comment by xjohnxjohn** [comment]
@alemelis I have 2 classifiers (age and gender), age have 103 classes (1-104 years) and gender  have 2 classes (Male and Female). When I test the predict, it show only Female and Male. What i am doing wrong?
my cfg:
...
[convolutional]
filters=103
size=1
stride=1
pad=1
activation=linear

[avgpool]

[softmax]
groups=1
classes=103


[route]
layers=22

[convolutional]
filters=2
size=1
stride=1
pad=1
activation=linear

[avgpool]

[softmax]
groups=1
classes=2


[route]
layers=25,29

**Comment by alemelis** [comment]
> @alemelis I have 2 classifiers (age and gender), age have 103 classes (1-103 years) and gender have 2 classes (Male and Female). When I test the predict, it show only Female and Male. What i am doing wrong?
> my cfg:
> ...
> [convolutional]
> filters=103
> size=1
> stride=1
> pad=1
> activation=linear
> 
> [avgpool]
> 
> [softmax]
> groups=1
> classes=103
> 
> [route]
> layers=22
> 
> [convolutional]
> filters=2
> size=1
> stride=1
> pad=1
> activation=linear
> 
> [avgpool]
> 
> [softmax]
> groups=1
> classes=2
> 
> [route]
> layers=25,29

hi @xjohnxjohn 
the second softmax layer is missing the `start_idx` parameter, it should be

```
[softmax]
groups=1
classes=2
start_idx=103
```

**Comment by xjohnxjohn** [comment]
@alemelis After i changed from ur recommendation, It show the same results (Female and Male). I tried to use the lastest repo and retraining, results are the same.

**Comment by alemelis** [comment]
> @alemelis After i changed from ur recommendation, It show the same results (Female and Male). I tried to use the lastest repo and retraining, results are the same.

you may be re-routing the layers in the wrong way, make sure 
```
[route]
layers=25,29
```
actually makes sense for your network

also, bear in mind that the master version is possibly not compatible with this branch anymore (there are conflicts)

**Comment by ManivannanMurugavel** [comment]
> @alemelis
> Thanks for your response.
> 
> I have already double-checked the .data, config and labels file, and everything is as per what you have mentioned.
> 
> However, I still get the below:
> 
> * model load information -
> 
> Too many or too few labels: 2, /path/to/image
> label 0: name1
> label1: name2

Hi @Germminate ,

The problem is, i will show my scenario,

"strstr" is used to check the substring exists or not.

I have to start the training with two classes and the class names is "cat and dog"
my path is "/home/manivannan/darknet_apr_22/classification/data/image113_dog.jpg"

The all classes is checking the substring of path.

So my scenario, the cat and dog is substring of the path.

But the path is mentioned the dog class.

Checking substring:
Cat:
/home/manivannan/darknet_apr_22/classifi**cat**ion/data/image113_dog.jpg
Dog:
/home/manivannan/darknet_apr_22/classification/data/image113_**dog**.jpg

Conclusion:
We have maintain the unique name of the path... The path is doesn't contain the substring of all class names.

Thanks,
